File: python/mnm/value.py
from . import _ffi
from . import context
from ._ffi._tvm import _NodeBase
from .base import register_mnm_node


@register_mnm_node("mnm.value.Value")
class Value(_NodeBase):
    pass


@register_mnm_node("mnm.value.TensorValue")
class TensorValue(Value):

    @property
    def dltensor_handle(self):
        return self.tensor.handle

    @property
    def data(self):
        handle = self.dltensor_handle
        return handle.contents.data

    @property
    def ctx(self):
        handle = self.dltensor_handle
        ctx = handle.contents.ctx
        return context.Context(ctx.device_type, ctx.device_id)

    @property
    def ndim(self):
        handle = self.dltensor_handle
        return handle.contents.ndim

    @property
    def dtype(self):
        handle = self.dltensor_handle
        return str(handle.contents.dtype)

    @property
    def shape(self):
        handle = self.dltensor_handle
        ndim = self.ndim
        return tuple(handle.contents.shape[i] for i in range(ndim))

    @property
    def strides(self):
        handle = self.dltensor_handle
        ndim = self.ndim
        return tuple(handle.contents.strides[i] for i in range(ndim))

    @property
    def byte_offset(self):
        handle = self.dltensor_handle
        return handle.contents.byte_offset

    @staticmethod
    def assemble(shape, dtype, ctx, strides=None, data=None):
        return _ffi.value.AssembleTensorValue(ctx, dtype, shape, strides, data)

File: src/mnm/value.cc
#include <mnm/registry.h>
#include <mnm/rly.h>
#include <mnm/tensor.h>
#include <mnm/value.h>

#include "./shape_utils.h"

namespace mnm {
namespace value {

using mnm::rly::Array;
using mnm::rly::Integer;
using mnm::rly::make_node;
using mnm::rly::NodePtr;
using mnm::shape_utils::MakeShape;
using mnm::tensor::Tensor;
using mnm::types::Context;
using mnm::types::DType;

TensorValue TensorValue::Assemble(Context ctx,                   //
                                  DType dtype,                   //
                                  std::vector<int64_t> shape,    //
                                  std::vector<int64_t> strides,  //
                                  void* data) {
  NodePtr<TensorValueNode> n = make_node<TensorValueNode>();
  n->tensor = Tensor::make(ctx, dtype, shape, strides, data);
  return TensorValue(n);
}

TensorValue AssembleTensorValue(DLContext ctx,            //
                                DLDataType dtype,         //
                                Array<Integer> _shape,    //
                                Array<Integer> _strides,  //
                                void* data) {
  return TensorValue::Assemble(ctx, dtype, MakeShape(_shape), MakeShape(_strides), data);
}

MNM_REGISTER_GLOBAL("mnm.value.AssembleTensorValue").set_body_typed(AssembleTensorValue);

}  // namespace value
}  // namespace mnm
